src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -17,7 +17,7 @@
 namespace mlpack {
 namespace tree {
 
-CosineTree::CosineTree(const arma::mat& dataset) :
+CosineTree::CosineTree(arma::mat& dataset) :
```

Comment:
After [this](https://github.com/mlpack/mlpack/pull/2023#discussion_r330858718) discussion with @walragatver I changed `dataset` to a non-const member.

Reply:
Got it, thanks for the explanation.  I see that in order to make these copy constructors and operators work, we need to be able to "re-seat" the reference.  The way that you've done that is by using `std::move()`.  That works, but the only thing is, it causes a user to expect their data to be modified (since the parameter is not `const`).  Also, it makes serialization much harder when the object holds a reference.  (I can see that serialization isn't implemented for the cosine tree, and I bet this is exactly why.)

So, I wanted to ask if you'd be willing to improve the `CosineTree` class by changing the member `dataset` from a reference to a pointer (i.e. `arma::mat* dataset`).  Then it would be like `BinarySpaceTree` and the other tree classes, and it would be much easier to implement the copy and move constructors and operators.  However, there would be a little additional work to make sure that the pointer was being properly handled, and we'd need to add a destructor.

We could use the same ownership rules as `BinarySpaceTree`: if `parent == NULL`, then we are responsible for deleting `dataset`.

Let me know what you think. :)